### PR TITLE
Noobaa upgrade check

### PR DESCRIPTION
**Pull request by fbalak**

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -1313,6 +1318,15 @@ def finalizer():
 
 @pytest.fixture()
 def mcg_obj(request):
+    return mcg_obj_fixture(request)
+
+
+@pytest.fixture(scope='session')
+def mcg_obj_session(request):
```
Comment (ebenahar) [inline tests/conftest.py]: First, it is a good idea to instantiate MCG per session. This is described in https://github.com/red-hat-storage/ocs-ci/issues/1178

I'm not though if with the current MCG cluster members, we can instantiate a single MCG for the entire session, as some of them are specific for a region / cloud provider.

File: tests/conftest.py [hunk 2]
```diff
@@ -1382,6 +1414,267 @@ def nodes():
...
+def uploaded_objects_session(request, mcg_obj_session, awscli_pod_session):
+    return uploaded_objects_fixture(
+        request,
+        mcg_obj_session,
+        awscli_pod_session
+    )
+
+
+def uploaded_objects_fixture(request, mcg_obj, awscli_pod):
+    """
```
Comment (ebenahar) [inline tests/conftest.py]: Please add empty lines after each section in the docstring

File: tests/conftest.py [hunk 3]
```diff
@@ -1382,6 +1414,267 @@ def nodes():
...
+    Deletes all objects that were created as part of the test
```
Comment (ebenahar) [inline tests/conftest.py]: Fixture name and its description don't match

File: tests/conftest.py [hunk 4]
```diff
@@ -1382,6 +1414,267 @@ def nodes():
...
+
+
+@pytest.fixture(scope='session')
+def bucket_factory_session(request, mcg_obj_session):
+    return bucket_factory_fixture(request, mcg_obj_session)
+
+
+def bucket_factory_fixture(request, mcg_obj):
+    """
+    Create a bucket factory. Calling this fixture creates a new bucket(s).
```
Comment (ebenahar) [inline tests/conftest.py]: add empty lines here between docstring sections

File: tests/conftest.py (hunk 1 above)
Comment (fbalak) [reply tests/conftest.py]: I need to instantiate it to session or class as upgrade test cases are broken into more tests (before and after upgrade).

File: tests/conftest.py (hunk 2 above)
Comment (fbalak) [reply tests/conftest.py]: Done

File: tests/conftest.py (hunk 4 above)
Comment (fbalak) [reply tests/conftest.py]: Ok

File: tests/ecosystem/upgrade/test_noobaa.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        command=f'mkdir {LOCAL_TESTOBJS_DIR_PATH}'
+    )
+    test_objects = boto3.resource('s3').Bucket(
+        constants.TEST_FILES_BUCKET
+    ).objects.all()
+
+    for obj in test_objects:
+        logger.info(f'Downloading {obj.key} from AWS test bucket')
+        awscli_pod_session.exec_cmd_on_pod(
+            command=f'sh -c "'
```
Comment (petr-balogh) [inline tests/ecosystem/upgrade/test_noobaa.py]: what about:
```python
            command=(
                f'sh -c "wget -P {LOCAL_TESTOBJS_DIR_PATH} '
                f'https://{constants.TEST_FILES_BUCKET}.s3.amazonaws.com/{obj.key}"'
            )
```

File: ocs_ci/ocs/resources/mcg_bucket.py [hunk 6]
```diff
@@ -33,6 +33,18 @@ def __eq__(self, other):
         elif type(other) == MCGBucket:
             return self.name == other.name
 
+    @property
+    def phase(self):
```
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg_bucket.py]: I think that all `obc` objects should have attribute `phase` no matter how they are created. But confirmation from @MeridianExplorer would be beneficial.

File: ocs_ci/ocs/resources/mcg.py [hunk 7]
```diff
@@ -598,12 +605,14 @@ def _check_state():
             sysinfo = self.send_rpc_query('system_api', 'read_system', params={}).json()['reply']
             for pool in sysinfo.get('pools'):
                 if pool.get('name') == backingstore_name:
-                    if pool.get('mode') == desired_state:
+                    current_state = pool.get('mode')
+                    logger.info(f'Current state of backingstore {backingstore_name} is {current_state}')
```
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg.py]: done

File: ocs_ci/ocs/resources/mcg.py [hunk 8]
```diff
@@ -615,6 +624,6 @@ def _check_state():
                     )
         except TimeoutExpiredError:
             logger.error(
-                'The BackingStore did not reach the desired state within the time limit.'
+                f'The BackingStore did not reach the desired state {desired_state} within the time limit.'
```
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg.py]: done

File: ocs_ci/ocs/resources/mcg_bucket.py [hunk 9]
```diff
@@ -33,6 +33,18 @@ def __eq__(self, other):
...
+        """
+        Returns phase of bucket claim
+
+        Returns:
+            str: OBC phase
+        """
+        return OCP(kind='obc', namespace=self.mcg.namespace).get(
+            resource_name=self.name
+        ).get('status').get('phase')
```
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg_bucket.py]: I agree, done

File: tests/conftest.py [hunk 10]
```diff
@@ -1404,6 +1439,311 @@ def nodes():
...
+
+
+def uploaded_objects_fixture(
+    request,
+    mcg_obj,
+    awscli_pod,
+    verify_rgw_restart_count
+):
+    """
+    Deletes all objects that were created as part of the test
```
[the remark this answers is not in the document]
Comment (fbalak) [reply tests/conftest.py]: done

File: ocs_ci/ocs/resources/mcg_bucket.py [hunk 11]
```diff
@@ -33,6 +33,18 @@ def __eq__(self, other):
...
             return self.name == other.name
 
+    @property
+    def phase(self):
+        """
+        Returns phase of bucket claim
+
+        Returns:
+            str: OBC phase
+        """
```
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg_bucket.py]: done

File: tests/conftest.py [hunk 12]
```diff
@@ -1404,6 +1439,311 @@ def nodes():
...
+@pytest.fixture(scope='session')
+def bucket_factory_session(request, mcg_obj_session):
+    return bucket_factory_fixture(request, mcg_obj_session)
+
+
+def bucket_factory_fixture(request, mcg_obj):
+    """
+    Create a bucket factory. Calling this fixture creates a new bucket(s).
+    For a custom amount, provide the 'amount' parameter.
+    Args:
```
[the remark this answers is not in the document]
Comment (fbalak) [reply tests/conftest.py]: done

File: tests/conftest.py [hunk 13]
```diff
@@ -1404,6 +1439,311 @@ def nodes():
...
+    bucketMap = {
+        's3': S3Bucket,
+        'oc': OCBucket,
+        'cli': CLIBucket
+    }
+
+    def _create_buckets(amount=1, interface='S3', *args, **kwargs):
+        """
+        Creates and deletes all buckets that were created as part of the test
+        Args:
```
[the remark this answers is not in the document]
Comment (fbalak) [reply tests/conftest.py]: done

File: tests/ecosystem/upgrade/test_noobaa.py (hunk 5 above)
Comment (fbalak) [reply tests/ecosystem/upgrade/test_noobaa.py]: sure

**Comment by fbalak** [comment]
Verification job: https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/5142/
Tests `test_fill_bucket` and `test_noobaa_postupgrade` passed but teardown failed as deletion of backing-store-s3bucket-testbs-311472a619e04e06a91276da64d50a6a timed out.

File: ocs_ci/ocs/resources/mcg.py [hunk 14]
```diff
@@ -582,39 +582,56 @@ def _check_mirroring():
             )
             assert False
 
-    def check_backingstore_state(self, backingstore_name, desired_state):
+    def check_backingstore_state(
+        self,
+        backingstore_name,
+        desired_state,
+        timeout=180
```
Comment (MeridianExplorer) [inline ocs_ci/ocs/resources/mcg.py]: Great change!
I recommend bringing the default timeout up to 600 seconds; Sometimes, the change takes a while to kick into action on the AWS side, and thus on the NooBaa side too

File: ocs_ci/ocs/resources/mcg_bucket.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MeridianExplorer) [reply ocs_ci/ocs/resources/mcg_bucket.py]: Here's the thing;
There are two "control planes" in the case of NooBaa. One is Kubernetes, and one is NooBaa/S3.
Whenever a user creates an S3 bucket using the S3 endpoint and API, it creates an OB. OBs are not accessible from Kubernetes, and it's impossible to see their status from there. The only way to check them is via the NooBaa UI, or RPC calls.
However, when a user creates an OBC using the CLI or a YAML, the OBC then subsequently creates an OB - which shows the phase.

Thus, if we'll want the phase of an OB, we'll have to retrieve it with RPC calls - which means S3Bucket's phase will be retrieved via RPC, and OCBucket+CLIBucket can use this code here.

File: tests/conftest.py (hunk 1 above)
Comment (MeridianExplorer) [reply tests/conftest.py]: Commented on the issue and described the situation - I think it's a subject for another PR

File: ocs_ci/ocs/resources/mcg.py (hunk 14 above)
Comment (fbalak) [reply ocs_ci/ocs/resources/mcg.py]: Thank you, updated.

**Comment by fbalak** [comment]
Another PR verification job:
https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/5627/
Tests from `tests/ecosystem/upgrade/test_noobaa.py` passed, upgrade itself timed out.

**Pull request merged**